Add validation TS docs (#1457)
encoredev/encore: docs/ts/primitives/validation.mdx

```ts
interface Request {
  value: string | number | boolean;
}

export const myEndpoint = api(
  { expose: true, method: "POST", path: "/body" },
  async (req: Request) => {
    // req.value is a string, number, or boolean
  },
);
```

### Reference schema

```ts
interface Schema {
  str: string; // String
  int: number; // Number
  list: number[]; // Array of numbers
  listOfTypes: (number | string )[]; // Array multiple types
  nullable: number | null; // Nullable
  maybe?: string; // Optional
  multiple: boolean | number | string | { name: string }; // Union
  enum: "John" | "Foo"; // Enum
}
```

<Callout type="info">

  For other types of validation, like checking the format of an email address or the length of a string, you need
  to implement those yourself inside the endpoint handler. More types of built-in validation will be added in the future.

</Callout>

## Body

By default, the data is parsed as a JSON body for incoming requests:

```ts
interface Request {
  name: string; // Parsed from the JSON body
}

export const myEndpoint = api<Request, Response>(
  { expose: true, method: "POST", path: "/body" },
  async (req) => {
    // req.name is a string
  },
);
```

> Here, `name` is a required field in the request body. If the request body is missing the `name` field, Encore will return a 400 Bad Request response.

Reviewer: ```suggestion
Here, `name` is a required field in the request body. If the request body is missing the `name` field, Encore will return a `400` Bad Request response.
```

```ts
import { Header, Query, api } from "encore.dev/api";

interface Request {
  // Required path parameter. Parsed from the request URL.
  id: number;

  // Optional query parameter. Parsed from the request URL.
  limit?: Query<number>;

  // Custom header that must be set. Parsed from the HTTP headers.
  myHeader: Header<"X-My-Header">;

  // Required enum. Parsed from the request body.
  type: "sprocket" | "widget";
}

export const myEndpoint = api(
  { expose: true, method: "POST", path: "/user/:id" },
  async ({ id, limit, myHeader, type }: Request) => {
    // ...
  },
);
```

## Errors

If the validation is not successful, Encore will return a `400` Bad Request response with a JSON body that contains the error message:

```output
HTTP/1.1 400 Bad Request

{
  "code": "invalid_argument",
  "message": "unable to decode request body",
  "internal_message": "Error(\"missing field name\", line: 1, column: 18)"
}
```

## Response

Encore.ts will not perform runtime validation for response data, but you will get compilation errors if you try to return a value that does not match the expected response type.

### Reusing the request type as the response type

> It is often the case that you want to return the same data type that you received in the request. In this case, you can reuse the request type as the response type:

Reviewer: ```suggestion
You often want to return the same data type that you received in a request. In this case, you can reuse the request type as the response type:
```

  body: string; // this will be sent as part of the JSON body
}
// A simple API endpoint that echoes the data back.
export const echo = api(
  { method: "POST", path: "/echo" },
  async (params: Data): Promise<Data> => {
    return params; // echo the data back
  },
);
```

This API endpoint expects incoming requests to look like this:
```output
POST /echo?query=hello HTTP/1.1
Content-Type: application/json
X-Header: this is a header
{
   "body": "a body",
}
```

For HTTP responses the `Query<string>` type is considered to be part of the JSON response body, since query strings only make sense for incoming requests.
Responses returned from this endpoint will be serialized as a HTTP response to looks like this:

```output
HTTP/1.1 200 OK
Content-Type: application/json
X-Header: this is a header
{
   "query": "hello",
   "body": "a body",
}
```

## Under the hood

Encore.ts parses your source code to understand the request and response schema that each API endpoint expects, including things like HTTP headers, query parameters, and so on. The schemas are then processed, optimized, and stored as a Protobuf file.

> When the Encore Rust Runtime starts up, it reads this Protobuf file and pre-computes a request decoder and response encoder, optimized for each API endpoint, using the exact type definition each API endpoint expects. In fact, Encore.ts even handles request validation directly in Rust, ensuring invalid requests never have to even touch the JS layer, mitigating many denial of service attacks.
> 
> Encore’s understanding of the request schema also proves beneficial from a performance perspective. JavaScript runtimes like Deno and Bun use a similar architecture to that of Encore's Rust-based runtime (in fact, Deno also uses Rust+Tokio+Hyper), but lack Encore’s understanding of the request schema. As a result, they need to hand over the un-processed HTTP requests to the single-threaded JavaScript engine for execution.
> 
> Encore.ts, on the other hand, handles much more of the request processing inside Rust, and only hands over the decoded request objects. By handling much more of the request life-cycle in multi-threaded Rust, the JavaScript event-loop is freed up to focus on executing application business logic instead of parsing HTTP requests, yielding an even greater performance boost.

Reviewer: ```suggestion
When the Encore.ts Rust runtime starts up, it reads the Protobuf file and pre-computes a request decoder and response encoder, optimized for each API endpoint, using the exact type definition each API endpoint expects. In fact, Encore.ts even handles request validation directly in Rust, ensuring invalid requests never have to even touch the JavaScript layer, mitigating many denial of service attacks.

Encore’s understanding of the request schema also improves performance. JavaScript runtimes like Deno and Bun use a similar architecture (in fact, Deno also uses Rust+Tokio+Hyper), but lack Encore’s understanding of the request schema. As a result, they need to hand over the un-processed HTTP requests to the single-threaded JavaScript engine for execution.

Encore.ts, on the other hand, handles much more of the request processing inside Rust, and only hands over the decoded request objects. By handling much more of the request life-cycle in multi-threaded Rust, the JavaScript event-loop is freed up to focus on executing application business logic instead of parsing HTTP requests, resulting in a significant performance improvement.
```

> 
> 

Reviewer: ```suggestion

```
